flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultPreferredLocationsRetriever.java

```diff
@@ -84,11 +86,24 @@ private CompletableFuture<Collection<TaskManagerLocation>> getPreferredLocations
         CompletableFuture<Collection<TaskManagerLocation>> preferredLocations =
                 CompletableFuture.completedFuture(Collections.emptyList());
 
-        final Collection<Collection<ExecutionVertexID>> allProducers =
-                inputsLocationsRetriever.getConsumedResultPartitionsProducers(executionVertexId);
-        for (Collection<ExecutionVertexID> producers : allProducers) {
+        final Collection<ConsumedPartitionGroup> consumedPartitionGroups =
+                inputsLocationsRetriever.getConsumedPartitionGroups(executionVertexId);
+        for (ConsumedPartitionGroup consumedPartitionGroup : consumedPartitionGroups) {
+            // Ignore the location of a consumed partition group if it has too many distinct
+            // consumers compared to the consumed partition group size. This is to avoid tasks
+            // unevenly distributed on nodes when running batch jobs or running jobs in
+            // session/standalone mode.
+            if ((double) consumedPartitionGroup.getConsumerVertexGroup().size()
+                            / consumedPartitionGroup.size()
+                    > MAX_DISTINCT_LOCATIONS_TO_CONSIDER) {
```

Comment:
Maybe introduce a new static field `MAX_DISTINCT_CONSUMERS_TO_CONSIDER` ?  The `MAX_DISTINCT_LOCATIONS_TO_CONSIDER` does not quite match the meaning here.